File: bindings/python/python/auths/protocol.py
from __future__ import annotations

import asyncio as _asyncio
import ssl as _ssl
import struct as _struct
import time as _time
import urllib.error as _urlerror
import urllib.parse as _urlparse
import urllib.request as _urlrequest
from dataclasses import dataclass as _dataclass
from datetime import timedelta as _timedelta
from typing import Any as _Any, Literal as _Literal, Optional as _Optional, Protocol as _Protocol, cast as _cast

from ._public import auths_error as _auths_error, error_info as _error_info, runtime_info as _runtime_info
from .verify import AuthorizedVerification, UnsuccessfulVerification, VerificationInput, VerificationMetrics, VerificationResult

_MEDIA: _Literal["application/vnd.auths.remote-verification.v1+cbor"] = "application/vnd.auths.remote-verification.v1+cbor"
_TOKEN = object()


def _head(major: int, value: int) -> bytes:
    if value < 24: return bytes([(major << 5) | value])
    if value <= 0xFF: return bytes([(major << 5) | 24, value])
    if value <= 0xFFFF: return bytes([(major << 5) | 25]) + _struct.pack(">H", value)
    if value <= 0xFFFFFFFF: return bytes([(major << 5) | 26]) + _struct.pack(">I", value)
    return bytes([(major << 5) | 27]) + _struct.pack(">Q", value)


def _encode(value: _Any) -> bytes:
    if value is None: return b"\xf6"
    if isinstance(value, bool): return b"\xf5" if value else b"\xf4"
    if isinstance(value, int) and value >= 0: return _head(0, value)
    if isinstance(value, bytes): return _head(2, len(value)) + value
    if isinstance(value, str):
        raw = value.encode(); return _head(3, len(raw)) + raw
    if isinstance(value, (list, tuple)): return _head(4, len(value)) + b"".join(_encode(v) for v in value)
    if isinstance(value, dict):
        items = sorted(value.items(), key=lambda item: _encode(item[0]))
        return _head(5, len(items)) + b"".join(_encode(k) + _encode(v) for k, v in items)
    raise TypeError("unsupported CBOR value")


def _decode(data: bytes) -> _Any:
    view = memoryview(data)
    def take(offset: int) -> tuple[_Any, int]:
        if offset >= len(view): raise ValueError
        initial = view[offset]; offset += 1; major, additional = initial >> 5, initial & 31
        if additional < 24: length = additional
        elif additional == 24: length = view[offset]; offset += 1
        elif additional == 25: length = _struct.unpack_from(">H", view, offset)[0]; offset += 2
        elif additional == 26: length = _struct.unpack_from(">I", view, offset)[0]; offset += 4
        elif additional == 27: length = _struct.unpack_from(">Q", view, offset)[0]; offset += 8
        elif major == 7 and additional in (20, 21, 22): return ({20: False, 21: True, 22: None}[additional], offset)
        else: raise ValueError
        if major == 0: return length, offset
        if major in (2, 3):
            end = offset + length
            if end > len(view): raise ValueError
            raw = bytes(view[offset:end]); return (raw if major == 2 else raw.decode("utf-8"), end)
        if major == 4:
            values: _Any = []
            for _ in range(length): value, offset = take(offset); values.append(value)
            return values, offset
        if major == 5:
            values = {}
            for _ in range(length):
                key, offset = take(offset); value, offset = take(offset)
                if key in values: raise ValueError
                values[key] = value
            return values, offset
        raise ValueError
    value, end = take(0)
    if end != len(view): raise ValueError
    return value


@_dataclass(frozen=True)
class TransportRequest:
    url: str
    method: _Literal["POST"]
    media_type: _Literal["application/vnd.auths.remote-verification.v1+cbor"]
    accept: _Literal["application/vnd.auths.remote-verification.v1+cbor"]
    body: bytes
    deadline_unix_ms: int
    maximum_response_bytes: int


@_dataclass(frozen=True)
class TransportResponse:
    status: int
    media_type: str
    body: bytes


class BoundedTransport(_Protocol):
    @property
    def contract(self) -> _Literal["bounded-byte-transport/2"]: ...
    async def send(self, request: TransportRequest) -> TransportResponse: ...
    async def aclose(self) -> None: ...


class _HttpTransport:
    contract: _Literal["bounded-byte-transport/2"] = "bounded-byte-transport/2"
    def __init__(self, token: str) -> None: self._token = token
    async def send(self, request: TransportRequest) -> TransportResponse:
        remaining = max(0.001, (request.deadline_unix_ms - int(_time.time() * 1000)) / 1000)
        value = _urlrequest.Request(request.url, data=request.body, method="POST", headers={"Authorization": "Bearer " + self._token, "Content-Type": request.media_type, "Accept": request.accept, "Auths-Error-Registry-SHA256": _runtime_info().error_registry_digest})
        def send() -> TransportResponse:
            opener = _urlrequest.build_opener(_urlrequest.HTTPSHandler(context=_ssl.create_default_context()), _NoRedirect())
            with opener.open(value, timeout=remaining) as response:
                body = response.read(request.maximum_response_bytes + 1)
                if len(body) > request.maximum_response_bytes: raise ValueError("response over limit")
                return TransportResponse(response.status, response.headers.get_content_type(), body)
        return await _asyncio.to_thread(send)
    async def aclose(self) -> None: return None


class _NoRedirect(_urlrequest.HTTPRedirectHandler):
    def redirect_request(self, req: _Any, fp: _Any, code: int, msg: str, headers: _Any, newurl: str) -> None:
        raise _urlerror.HTTPError(req.full_url, code, "redirect refused", headers, fp)


class RemoteVerifier:
    def __init__(self, token: object, endpoint: str, transport: BoundedTransport, owns: bool, timeout: _timedelta, maximum: int, insecure: bool) -> None:
        if token is not _TOKEN: raise TypeError("RemoteVerifier is sealed")
        parsed = _urlparse.urlsplit(endpoint)
        loopback = parsed.hostname in ("127.0.0.1", "::1")
        if (parsed.scheme != "https" and not (insecure and parsed.scheme == "http" and loopback)) or not parsed.netloc or parsed.username or parsed.path not in ("", "/") or parsed.query or parsed.fragment: raise ValueError("endpoint must be an HTTPS origin")
        seconds = timeout.total_seconds()
        if not 0.001 <= seconds <= 300 or seconds * 1000 != int(seconds * 1000): raise ValueError("timeout outside bounds")
        if not 1024 <= maximum <= 16 * 1024 * 1024: raise ValueError("maximum response outside bounds")
        self._url, self._transport, self._owns, self._timeout, self._maximum = parsed.scheme + "://" + parsed.netloc + "/v2/verification/authorize", transport, owns, seconds, maximum
        self._state = "new"
    async def __aenter__(self) -> "RemoteVerifier":
        if self._state != "new": raise RuntimeError("auths client is not open")
        self._state = "open"; return self
    async def __aexit__(self, *exc: object) -> None: await self.aclose()
    async def aclose(self) -> None:
        if self._state == "closed": return
        self._state = "closed"
        if self._owns: await self._transport.aclose()
    async def verify(self, input: VerificationInput, /) -> VerificationResult:
        if self._state != "open": raise RuntimeError("auths client is not open")
        correlation = f"auths-{_time.time_ns():x}"
        digest = bytes.fromhex(_runtime_info().error_registry_digest)
        body = _encode({0: 1, 1: input.proof, 2: input.action, 3: input.trusted_context, 4: correlation, 5: digest})
        request = TransportRequest(self._url, "POST", _MEDIA, _MEDIA, body, int((_time.time() + self._timeout) * 1000), self._maximum)
        try: response = await _asyncio.wait_for(self._transport.send(request), self._timeout)
        except _asyncio.TimeoutError: raise _auths_error("remote.timeout")
        except Exception as exc: raise _auths_error("remote.transport-unavailable", summary=str(exc)[:256])
        if response.status != 200 or response.media_type.split(";", 1)[0].strip() != _MEDIA: raise _auths_error("remote.response-malformed")
        try:
            raw = _decode(response.body)
            if raw[0] != 1 or raw[4] != correlation or raw[10] != digest: raise ValueError
            kinds, stages = ("authorized", "denied", "indeterminate"), ("decode", "resolve", "principal-control", "authority", "complete")
            kind = kinds[raw[1]]
            stage = _cast(_Literal["decode", "resolve", "principal-control", "authority", "complete"], stages[raw[3]])
            metrics_raw = raw[5]
            metrics = VerificationMetrics(*(int(metrics_raw[index]) for index in range(7)))
            code = str(raw[2])
            required = None if raw[6] is None else bytes(raw[6])
            executed = bytes(raw[7])
            decision = bytes(raw[8])
            if kind == "authorized":
                return AuthorizedVerification("authorized", code, stage, correlation, metrics, required, executed, decision)
            negative_kind: _Literal["denied", "indeterminate"] = "denied" if kind == "denied" else "indeterminate"
            return UnsuccessfulVerification(negative_kind, code, stage, correlation, metrics, required, executed, decision, _error_info(raw[9].get("code", "core.authorization-indeterminate")))
        except Exception: raise _auths_error("remote.response-malformed")


def connect_remote_verifier(*, endpoint: str, access_token: str, timeout: _timedelta = _timedelta(seconds=30), maximum_response_bytes: int = 8 * 1024 * 1024, allow_insecure_loopback: bool = False) -> RemoteVerifier:
    if not access_token or len(access_token.encode()) > 8192: raise ValueError("invalid access token")
    return RemoteVerifier(_TOKEN, endpoint, _HttpTransport(access_token), True, timeout, maximum_response_bytes, allow_insecure_loopback)


def remote_verifier_from_transport(endpoint: str, transport: BoundedTransport, /, *, owns_transport: bool = False, timeout: _timedelta = _timedelta(seconds=30), maximum_response_bytes: int = 8 * 1024 * 1024, allow_insecure_loopback: bool = False) -> RemoteVerifier:
    if transport.contract != "bounded-byte-transport/2": raise ValueError("unsupported bounded transport contract")
    return RemoteVerifier(_TOKEN, endpoint, transport, owns_transport, timeout, maximum_response_bytes, allow_insecure_loopback)


__all__ = ["TransportRequest", "TransportResponse", "BoundedTransport", "RemoteVerifier", "connect_remote_verifier", "remote_verifier_from_transport"]

File: bindings/typescript/src/protocol.ts
import type { VerificationInput, VerificationResult } from "./verify.js";
import { authsErrorFromIssue } from "./product-errors.js";
import { issue } from "./internal/issues.js";
import { runtimeInfo } from "./index.js";
import { decodeDeterministic, encodeDeterministic } from "./internal/cbor.js";

const mediaType = "application/vnd.auths.remote-verification.v1+cbor" as const;

export interface BoundedTransportRequest {
  readonly url: URL;
  readonly method: "POST";
  readonly mediaType: typeof mediaType;
  readonly accept: typeof mediaType;
  readonly body: Uint8Array;
  readonly deadlineUnixMs: number;
  readonly signal: AbortSignal;
  readonly maximumResponseBytes: number;
}
export interface BoundedTransportResponse { readonly status: number; readonly mediaType: string; readonly body: Uint8Array }
export interface BoundedTransport extends AsyncDisposable {
  readonly contract: "bounded-byte-transport/2";
  send(request: BoundedTransportRequest): Promise<BoundedTransportResponse>;
  close(): Promise<void>;
  [Symbol.asyncDispose](): Promise<void>;
}
export interface RemoteVerifier extends AsyncDisposable {
  verify(input: VerificationInput & Readonly<{ signal?: AbortSignal }>): Promise<VerificationResult>;
  close(): Promise<void>;
  [Symbol.asyncDispose](): Promise<void>;
}
interface RemoteVerifierCommonOptions {
  readonly endpoint: string | URL;
  readonly timeoutMs?: number;
  readonly maximumResponseBytes?: number;
  readonly allowInsecureLoopback?: boolean;
}
export type RemoteVerifierOptions = RemoteVerifierCommonOptions & (
  | Readonly<{ accessToken: string; transport?: never; transportOwnership?: never }>
  | Readonly<{ accessToken?: never; transport: BoundedTransport; transportOwnership?: "borrowed" | "owned" }>
);

export async function connectRemoteVerifier(options: RemoteVerifierOptions): Promise<RemoteVerifier> {
  const endpoint = parseEndpoint(options.endpoint, options.allowInsecureLoopback ?? false);
  const timeoutMs = boundedInteger(options.timeoutMs ?? 30_000, 1, 300_000, "timeout");
  const maximumResponseBytes = boundedInteger(options.maximumResponseBytes ?? 8 * 1024 * 1024, 1024, 16 * 1024 * 1024, "response limit");
  const injected = "transport" in options && options.transport !== undefined;
  const transport = injected ? options.transport : builtinTransport(options.accessToken);
  const owned = injected ? options.transportOwnership === "owned" : true;
  return new RemoteVerifierImpl(endpoint, timeoutMs, maximumResponseBytes, transport, owned);
}

class RemoteVerifierImpl implements RemoteVerifier {
  #state: "open" | "closing" | "closed" = "open";
  readonly #endpoint: URL;
  readonly #timeoutMs: number;
  readonly #maximumResponseBytes: number;
  readonly #transport: BoundedTransport;
  readonly #owned: boolean;

  constructor(endpoint: URL, timeoutMs: number, maximumResponseBytes: number, transport: BoundedTransport, owned: boolean) {
    this.#endpoint = endpoint;
    this.#timeoutMs = timeoutMs;
    this.#maximumResponseBytes = maximumResponseBytes;
    this.#transport = transport;
    this.#owned = owned;
  }

  async verify(input: VerificationInput & Readonly<{ signal?: AbortSignal }>): Promise<VerificationResult> {
    if (this.#state !== "open") throw new TypeError("Auths remote verifier is not open");
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(new DOMException("Auths verification timed out", "TimeoutError")), this.#timeoutMs);
    const abort = () => controller.abort(input.signal?.reason);
    input.signal?.addEventListener("abort", abort, { once: true });
    try {
      const correlationId = `auths-${crypto.randomUUID()}`;
      const registryDigest = (await runtimeInfo()).errorRegistryDigest;
      const body = encodeRequest(input, correlationId, registryDigest);
      const response = await this.#transport.send({
        url: new URL("/v2/verification/authorize", this.#endpoint),
        method: "POST",
        mediaType,
        accept: mediaType,
        body,
        deadlineUnixMs: Date.now() + this.#timeoutMs,
        signal: controller.signal,
        maximumResponseBytes: this.#maximumResponseBytes,
      });
      if (response.status !== 200 || response.mediaType.split(";", 1)[0]?.trim() !== mediaType || response.body.length > this.#maximumResponseBytes) {
        throw authsErrorFromIssue(issue("remote.response-malformed", { causes: ["invalid-response"] }));
      }
      return decodeResult(response.body, correlationId, registryDigest);
    } catch (error) {
      if (error instanceof Error && error.name === "AuthsError") throw error;
      if (controller.signal.aborted) throw authsErrorFromIssue(issue("remote.timeout", { causes: ["timeout"] }));
      throw authsErrorFromIssue(issue("remote.transport-unavailable", { causes: ["unavailable"] }));
    } finally {
      clearTimeout(timer);
      input.signal?.removeEventListener("abort", abort);
    }
  }

  async close(): Promise<void> {
    if (this.#state === "closed") return;
    this.#state = "closing";
    if (this.#owned) await this.#transport.close();
    this.#state = "closed";
  }
  async [Symbol.asyncDispose](): Promise<void> { await this.close(); }
}

function builtinTransport(accessToken: string): BoundedTransport {
  if (typeof accessToken !== "string" || accessToken.length < 1 || accessToken.length > 8192 || /[\u0000-\u001f\u007f]/u.test(accessToken)) throw new TypeError("invalid Auths access token");
  let closed = false;
  return {
    contract: "bounded-byte-transport/2",
    async send(request) {
      if (closed) throw new TypeError("transport is closed");
      const response = await fetch(request.url, {
        method: request.method,
        redirect: "error",
        headers: { "authorization": `Bearer ${accessToken}`, "content-type": request.mediaType, "accept": request.accept, "auths-error-registry-sha256": (await runtimeInfo()).errorRegistryDigest },
        body: request.body.slice().buffer as ArrayBuffer,
        signal: request.signal,
      });
      const declared = Number(response.headers.get("content-length") ?? "0");
      if (Number.isFinite(declared) && declared > request.maximumResponseBytes) throw new RangeError("remote response exceeds bound");
      const bytes = new Uint8Array(await response.arrayBuffer());
      if (bytes.length > request.maximumResponseBytes) throw new RangeError("remote response exceeds bound");
      return Object.freeze({ status: response.status, mediaType: response.headers.get("content-type") ?? "", body: bytes });
    },
    async close() { closed = true; },
    async [Symbol.asyncDispose]() { closed = true; },
  };
}

function parseEndpoint(value: string | URL, allowInsecureLoopback: boolean): URL {
  const url = new URL(value);
  if (url.username || url.password || url.search || url.hash || (url.pathname !== "/" && url.pathname !== "")) throw new TypeError("endpoint must be an origin");
  const loopback = url.hostname === "127.0.0.1" || url.hostname === "[::1]" || url.hostname === "localhost";
  if (url.protocol !== "https:" && !(allowInsecureLoopback && loopback && url.protocol === "http:")) throw new TypeError("Auths endpoint must use HTTPS");
  return new URL(url.origin);
}

function boundedInteger(value: number, minimum: number, maximum: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < minimum || value > maximum) throw new RangeError(`${name} is outside bounds`);
  return value;
}

function encodeRequest(input: VerificationInput, correlationId: string, registryDigest: string): Uint8Array {
  if (!(input.proof instanceof Uint8Array) || !(input.action instanceof Uint8Array) || !(input.trustedContext instanceof Uint8Array)) throw new TypeError("verification input must contain bytes");
  return encodeDeterministic(new Map<unknown, unknown>([
    [0, 1], [1, input.proof], [2, input.action], [3, input.trustedContext],
    [4, correlationId], [5, hexBytes(registryDigest)],
  ]));
}

function decodeResult(bytes: Uint8Array, correlationId: string, registryDigest: string): VerificationResult {
  const value = decodeDeterministic(bytes);
  if (!(value instanceof Map) || value.size !== 11 || value.get(0) !== 1 || value.get(4) !== correlationId || !equalBytes(asBytes(value.get(10)), hexBytes(registryDigest))) throw new TypeError("malformed remote verification result");
  const kinds = ["authorized", "denied", "indeterminate"] as const;
  const stages = ["decode", "resolve", "principal-control", "authority", "complete"] as const;
  const kind = kinds[asNumber(value.get(1))]; const stage = stages[asNumber(value.get(3))];
  const metrics = value.get(5);
  if (kind === undefined || stage === undefined || !(metrics instanceof Map)) throw new TypeError("malformed remote verification result");
  const common = {
    code: asString(value.get(2)), stage, correlationId,
    metrics: Object.freeze({ proofBytes: asBigInt(metrics.get(0)), actionBytes: asBigInt(metrics.get(1)), contextBytes: asBigInt(metrics.get(2)), objectCount: asBigInt(metrics.get(3)), planLeaves: asBigInt(metrics.get(4)), planDepth: asBigInt(metrics.get(5)), workUnits: asBigInt(metrics.get(6)) }),
    ...(value.get(6) === null ? {} : { requiredConfiguration: asBytes(value.get(6)) }),
    executedConfiguration: asBytes(value.get(7)), decisionBytes: asBytes(value.get(8)),
  };
  if (kind === "authorized") return Object.freeze({ ...common, kind: "authorized" as const });
  const envelope = value.get(9); const code = envelope instanceof Map && typeof envelope.get("code") === "string" ? envelope.get("code") as string : kind === "denied" ? "core.authorization-denied" : "core.authorization-indeterminate";
  if (kind === "denied") return Object.freeze({ ...common, kind: "denied" as const, issue: issue(code as never) }) as VerificationResult;
  return Object.freeze({ ...common, kind: "indeterminate" as const, issue: issue(code as never) }) as VerificationResult;
}

function hexBytes(value: string): Uint8Array {
  if (!/^[0-9a-f]{64}$/u.test(value)) throw new TypeError("invalid registry digest");
  return Uint8Array.from({ length: 32 }, (_, index) => Number.parseInt(value.slice(index * 2, index * 2 + 2), 16));
}
function asBytes(value: unknown): Uint8Array { if (!(value instanceof Uint8Array)) throw new TypeError("expected bytes"); return value; }
function asString(value: unknown): string { if (typeof value !== "string") throw new TypeError("expected string"); return value; }
function asNumber(value: unknown): number { if (typeof value !== "number" || !Number.isSafeInteger(value)) throw new TypeError("expected integer"); return value; }
function asBigInt(value: unknown): bigint { return typeof value === "bigint" ? value : BigInt(asNumber(value)); }
function equalBytes(left: Uint8Array, right: Uint8Array): boolean { return left.length === right.length && left.every((value, index) => value === right[index]); }
